### ISSUE TITLE
parentheses not showing up in comment

### ISSUE THREAD

**Issue by dtenenba** [Bioconductor/legacy.support.bioconductor.org#21]
Hi @ialbert ,

In a comment, Hervé  ( @hpages ) included the following (this is from looking at the editor in "source" mode):

```
Now at this point you could try <code>as.character(mcols(genes.ENSEMBL.MT)$entrez_id)</code>&nbsp;and it will succeed
```

However, after saving the comment, this rendered without the opening parentheses:

```
Now at this point you could try as.charactermcolsgenes.ENSEMBL.MT)$entrez_id) and it will succeed
```

Do you know what's going on there?

The comment in question is

https://support.bioconductor.org/p/64290/#64796

However, by the time you look at it you probably won't see this issue because @hpages found a workaround which is to put a space after the open parenthesis. 

Feel free to create a test post.

This could be an issue which is already fixed in ialbert/biostar-central; I haven't pulled from it in a while.
I should probably check with you before I do, I recall you saying something about needing to run some migrations if I do any pulls.

Thanks,
Dan

**Comment by ialbert** [comment]
this might be related to parsing html links - specifically autolinking. The
markdown library (third party library) has some features to recognize URLs
in text and automatically create  html links out of them for example foo.com
would become an html link.

Now the library originally was a little too eager and would recognize even
domain-less links and think of  foo.py as Paraguay domain and foo.pl as
Polish domain etc.

I remember having to work around some of these issues a lot. The
autolinking is very handy for simple domains but as it turns out it is very
complicated to get right.

For example here it is supposed to turn off the autolinking for links
without a domain:

https://github.com/Bioconductor/support.bioconductor.org/blob/master/biostar/apps/util/html.py#L81

but then there is a special case for parentheses since those from the point
of view of URL spec can be at the end of a URL in practice the last
parenthesis is always part of the text not the link.

Your example still passes through, I think it might because it is inside a
code tag that might execute differently.

Long story short I will look into it.

It is a bug a combination of features of the markdown parser probably in
conjunction with other extensions that I have added.

**Comment by dtenenba** [comment]
Thanks. But I would have imagined that inside a &lt;code&gt; tag, there is no markdown parsing at all, because that text is supposed to be verbatim, right?

**Comment by ialbert** [comment]
yes that would be my expectation too that markdown turns itself off when in
block but then looks like the parenthesis matching does not

**Comment by ialbert** [comment]
Actually I misspoke above I am using the bleach html sanitizer to linkify and not a markdown parser. 

http://bleach.readthedocs.org/en/latest/index.html

I now believe this to be an internal issue with the way linkifying works in bleach. I have opened an issue here:

https://github.com/jsocol/bleach/issues/146

I will also say that I have tried to implement this myself many times with various successes and I've come to believe that correctly linkifying links in text is a surprisingly and unexpectedly complicated task. 

In the meantime you have the following options. 
- Turn off autolinking altogether: comment out the line:

https://github.com/Bioconductor/support.bioconductor.org/blob/master/biostar/apps/util/html.py#L112
-  Allow the library to create links  to domain-like words `genes.ENSEMBL.MT` in a way that does not take the user to another page for example href='#'.  For that change line:

https://github.com/Bioconductor/support.bioconductor.org/blob/master/biostar/apps/util/html.py#L83

from `return None` to `attrs['href'] = '#'`
- Wait for changes upstream in the bleach library 

Alas this problem affects normal text as well. Domain like words with a leading parenthesis will lose the parenthesis.

**Comment by ialbert** [comment]
There is now a fix in the bleach library, the package update should not be long now

https://github.com/jsocol/bleach/issues/146#event-238419278

**Comment by dtenenba** [comment]
Great, thanks. Will you let me know when the update is available and how I should get it?
Thanks.

**Comment by ialbert** [comment]
Once the new version 1.4.2 you can install with

```
pip install --upgrade bleach
```

This will replace your current version with the updated one. It is not clear when that will be out the author seems to have started the changelog for it. Can't be too long.

On the other hand you can also apply the fix directly from the git repository like so:

```
pip install -e git+git@github.com:jsocol/bleach.git#egg=bleach
```

Now in this case you will have the library twice and this latter comes first in the import path so it will work. When upgrading it may take some work to update the right library.

To test that it worked run this:

```
from bleach import linkify
expect = '(test.py)'
print linkify(expect, callbacks=[lambda *a: None])
```

should print `(test.py)`
